src/gui_common/TweakedColourPicker.cs

```diff
@@ -0,0 +1,473 @@
+    {
+        if (colour.IsValidHtmlColor())
+        {
+            Color = new Color(colour);
+        }
+        else
+        {
+            htmlColourEdit.Text = Color.ToHtml();
+        }
+    }
+
+    /// <summary>
+    ///   Called when focus exited HtmlColourEdit.
+    /// </summary>
+    private void OnHtmlColourEditFocusExited()
+    {
+        OnHtmlColourEditEntered(htmlColourEdit.Text);
+        htmlColourEdit.Deselect();
+    }
+
+    private class TweakedColourPickerPreset : ColorRect
+    {
+        private readonly TweakedColourPicker owner;
+
+        public TweakedColourPickerPreset(TweakedColourPicker owner, Color colour)
+        {
+            this.owner = owner;
+            Color = colour;
+
+            // Init the GUI part of the ColorRect
+            MarginTop = MarginBottom = MarginLeft = MarginRight = 6.0f;
+            RectMinSize = new Vector2(20, 20);
+            SizeFlagsHorizontal = (int)SizeFlags.ShrinkEnd;
+            SizeFlagsVertical = (int)SizeFlags.ShrinkCenter;
+            UpdateTooltip();
+        }
+
+        public override void _Notification(int what)
+        {
+            if (what == NotificationTranslationChanged)
+                UpdateTooltip();
+
+            base._Notification(what);
+        }
+
+        public override void _GuiInput(InputEvent inputEvent)
+        {
+            if (inputEvent is InputEventMouseButton { Pressed: true } mouseEvent)
+            {
+                switch ((ButtonList)mouseEvent.ButtonIndex)
+                {
+                    case ButtonList.Left:
+                        GetTree().SetInputAsHandled();
+                        owner.Color = Color;
+                        break;
+                    case ButtonList.Right:
+                        if (!owner.PresetsEditable)
+                            break;
+
+                        GetTree().SetInputAsHandled();
+                        owner.ErasePreset(Color);
+                        break;
+                }
+            }
+
+            base._GuiInput(inputEvent);
+        }
+
+        private void UpdateTooltip()
+        {
+            HintTooltip = string.Format(CultureInfo.CurrentCulture,
+                TranslationServer.Translate("COLOUR_PICKER_PRESET_TOOLTIP"), Color.ToHtml());
+        }
+    }
+
+    private class PresetGroupStorage : IEnumerable<Color>
+    {
+        private readonly List<Color> colours;
+
+        public PresetGroupStorage(IEnumerable<Color> colours)
+        {
+            this.colours = colours.ToList();
+        }
+
+        public AddPresetDelegate AddPresetDelegate { get; set; }
+
+        public DeletePresetDelegate ErasePresetDelegate { get; set; }
+
+        public void AddPreset(Color colour)
+        {
+            if (colours.Contains(colour))
+                return;
+
+            colours.Add(colour);
+
+            // ?. is no used because we can make sure that at least one delegate is added when this get called.
```

Comment:
I would still prever to use the "proper" way here.

Reply:
Okay